### ISSUE TITLE
关于 2D RoPE implementation x, y pos 顺序的问题

### ISSUE THREAD

**Issue by yihaocs** [Meituan-AutoML/VisionLLaMA#4]
您好！

Amazing work！
关于2D RoPE implementation x_pos, y_pos 顺序我有一些疑问. 在 code: https://github.com/Meituan-AutoML/VisionLLaMA/blob/main/mmpretrain/mmpretrain/models/backbones/vision_llama.py#L34-L52 的实现中:
```python
x_pos = flat_patch_pos % W # N
x_pos = bias + x_pos * step
y_pos = flat_patch_pos // W # N
y_pos = bias + y_pos * step
```
假如W=5 (width), H=3 (Height)，那么坐标的变换应该是 
```
(0,0), (1, 0), (2, 0), (3, 0), (4,0),  (0,1), (1,1), (2,1)....... (4,2)
``` 
这里是把宽(width)的坐标放在了前面，高(height)的坐标放在了后面；然后:
```python
freqs_cis = torch.cat([x_cis.unsqueeze(dim=-1), y_cis.unsqueeze(dim=-1)], dim=-1) # N,Hc/4,2
```
那么相当于把width的变化放在了前面。

但是在传入block之前的x: https://github.com/Meituan-AutoML/VisionLLaMA/blob/main/dit/models.py#L442, 

https://github.com/pprp/timm/blob/master/timm/layers/patch_embed.py#L55:
```
def forward(self, x):
        B, C, H, W = x.shape
        _assert(H == self.img_size[0], f"Input image height ({H}) doesn't match model ({self.img_size[0]}).")
        _assert(W == self.img_size[1], f"Input image width ({W}) doesn't match model ({self.img_size[1]}).")
        x = self.proj(x)
        if self.flatten:
            x = x.flatten(2).transpose(1, 2)  # BCHW -> BNC
        x = self.norm(x)
        return x
```
这里flatten成length的顺序是 H X W, 那么讲道理应该把height的坐标放在前面，width的坐标放在后面，坐标变换的顺序应该是
```
(0,0), (0, 1), (0, 2), (0, 3), (0,4),  (1,0), (1,1), (1,2)....... (2,4)
``` 

正确的写法是否是这样子:
```python
def precompute_freqs_cis_2d_general(dim: int,  end: int = 1, step:int =1,  W=0, bias=0.0,  theta: float = 10000.0, scale=1.0, use_cls=False):
    if W == 0:
        W = int( end**0.5 )
    # assert  H * H == end
    flat_patch_pos = torch.arange(0 if not use_cls else -1, end) # N = end
    h_pos = flat_patch_pos // W # N
    h_pos = bias + x_pos * step
    w_pos = flat_patch_pos % W # N
    w_pos = bias + y_pos * step
    freqs = 1.0 / (theta ** (torch.arange(0, dim, 4)[: (dim // 4)].float() / dim)) # Hc/4
    h_freqs = torch.outer(h_pos, freqs).float() # N Hc/4
    w_freqs = torch.outer(w_pos, freqs).float() # N Hc/4
    h_cis = torch.polar(torch.ones_like(h_freqs), h_freqs)
    w_cis = torch.polar(torch.ones_like(w_freqs), w_freqs)
    freqs_cis = torch.cat([h_cis.unsqueeze(dim=-1), w_cis.unsqueeze(dim=-1)], dim=-1) # N,Hc/4,2
    freqs_cis = freqs_cis.reshape(end if not use_cls else end + 1, -1)
    # we need to think how to implement this for multi heads.
    # freqs_cis = torch.cat([h_cis, w_cis], dim=-1) # N, Hc/2
    return freqs_cis
```

我又仔细看了一遍Flexible DIT paper里面关于2D RoPE的推导，也是把height放在前面:

<img width="552" alt="Screenshot 2024-04-04 at 4 32 00 PM" src="https://github.com/Meituan-AutoML/VisionLLaMA/assets/93098586/176f4b4c-8f60-4c16-b894-ad9990939d88">

这个是比较符直觉的. 你们paper里面示意图也是这么画的:

<img width="397" alt="Screenshot 2024-04-04 at 4 33 44 PM" src="https://github.com/Meituan-AutoML/VisionLLaMA/assets/93098586/5b9e7e68-20a1-4f70-9c32-16ef0557c37e">

如果我miss了什么地方，请多多指教！

Looking forward to your Reply!

**Comment by cxxgtxy** [comment]
Thanks for your attention.
X (H,W) is flatten into HW, from top to down, from left to right. We make sure every dot has the right position. Just keeping the consistent form is ok.
In our axis, x is for w, y is for h.
